bsp/src/mill/bsp/BspWorker.scala

```diff
@@ -0,0 +1,80 @@
+package mill.bsp
+
+import mill.Agg
+import mill.api.{Ctx, PathRef, Result, internal}
+import mill.define.Task
+import mill.eval.Evaluator
+import mill.main.{BspServerHandle, BspServerResult}
+
+import java.io.{InputStream, PrintStream}
+import java.net.URL
+import scala.concurrent.Promise
+import scala.util.{Failure, Success, Try}
+
+@internal
+trait BspWorker {
+
+  def createBspConnection(
+      jobs: Int,
+      serverName: String
+  )(implicit ctx: Ctx): Unit
+
+  def startBspServer(
+      initialEvaluator: Option[Evaluator],
+      outStream: PrintStream,
+      errStream: PrintStream,
+      inStream: InputStream,
+      logDir: os.Path,
+      canReload: Boolean,
+      serverHandles: Seq[Promise[BspServerHandle]]
+  ): BspServerResult
+
+}
+
+@internal
+object BspWorker {
+
+  private[this] var worker: Option[BspWorker] = None
+
+  def apply(millCtx: Ctx.Workspace with Ctx.Home with Ctx.Log): Result[BspWorker] = {
+    worker match {
+      case Some(x) => Result.Success(x)
+      case None =>
+        // load extra classpath entries from file
+        val cpFile = millCtx.workspace / Constants.bspDir / s"${Constants.serverName}.resources"
+        if (!os.exists(cpFile)) return Result.Failure(
+          "You need to run `mill mill.bsp.BSP/install` before you can use the BSP server"
+        )
+
+        // TODO: if outdated, we could regenerate the resource file and re-load the worker
```

Comment:
This would be nice, detect the Mill version somehow and warn the user if there is a mismatch.

Reply:
Ahhh, I see we actually do this down below, so is this a leftover comment?

Reply:
No, it's still valid. We detect that we are outdated, but at this point we can't do anything about it, except logging or failing. Failing is too rude here, as we would force the user to understand and handle it. Logging is only helpful to understand the issue when we get access to the logs.

At this stage (firing up BSP server without having a completely initialized Evaluator) we can't trigger a re-build of the BSP connection file, as we can't use the evaluator and the task graph.

Reply:
One way to handle a detected outdated worker could be to:

* remember that fact in the BSP service
* Wait for an Evaluator
* Trigger a rebuild of the connection file
* Send an event to the client, that we need to re-start or (if possible) are going to restart the server